src/main.rs

```diff
@@ -34,321 +33,323 @@ extern crate xi_core_lib;
 extern crate xi_rpc;
 #[macro_use]
 extern crate xi_win_shell;
+extern crate xi_win_ui;
 
-mod dialog;
 mod edit_view;
 mod linecache;
 mod menus;
 mod rpc;
 mod textline;
 mod xi_thread;
 
-use std::any::Any;
-use std::cell::{RefCell, RefMut};
-use std::rc::Rc;
 use std::sync::{Arc, Mutex};
-
-use winapi::shared::windef::*;
+use std::collections::HashMap;
 
 use serde_json::Value;
 
 use edit_view::EditView;
 use menus::MenuEntries;
 use rpc::{Core, Handler};
-use xi_win_shell::util::Error;
-use dialog::{get_open_file_dialog_path, get_save_file_dialog_path};
 use xi_thread::start_xi_thread;
 
-use xi_win_shell::paint::PaintCtx;
-use xi_win_shell::win_main::{self, RunLoopHandle};
-use xi_win_shell::window::{Cursor, IdleHandle, MouseEvent, WindowBuilder,
-    WindowHandle, WinHandler};
+use xi_win_shell::win_main::{self};
+use xi_win_shell::window::{Cursor, IdleHandle, WindowBuilder};
 
-struct MainWinState {
-    edit_view: EditView,
-}
+use xi_win_ui::{UiMain, UiState};
+use xi_win_ui::Id;
+use xi_win_ui::{FileDialogOptions, FileDialogType};
 
-impl MainWinState {
-    fn new() -> MainWinState {
-        MainWinState {
-            edit_view: EditView::new(),
-        }
-    }
+use edit_view::EditViewCommands;
 
-    fn render(&mut self, p: &mut PaintCtx) {
-        self.edit_view.render(p);
-    }
-}
+type ViewId = String;
 
-struct MainWinHandler {
-    win: Rc<MainWin>,
+#[derive(Clone)]
+struct ViewState {
+    id: Id,
+    filename: Option<String>,
+    handle: Arc<Mutex<IdleHandle>>,
 }
 
-// Maybe combine all this, put as a single item inside a RefCell.
-pub struct MainWin {
-    core: RefCell<Core>,
-    handle: RefCell<WindowHandle>,
-    state: RefCell<MainWinState>,
+#[derive(Clone)]
+struct AppState {
+    focused: ViewId,
+    views: HashMap<ViewId, ViewState>,
 }
 
-impl MainWin {
-    fn new(core: Core, state: MainWinState) -> MainWin {
-        MainWin {
-            core: RefCell::new(core),
-            handle: Default::default(),
-            state: RefCell::new(state),
-        }
-    }
-
-    fn send_notification(&self, method: &str, params: &Value) {
-        self.core.borrow().send_notification(method, params);
-    }
-
-    // Note: caller can't be borrowing the state.
-    fn send_edit_cmd(&self, method: &str, params: &Value, view_id: &str) {
-        let edit_params = json!({
-            "method": method,
-            "params": params,
-            "view_id": view_id,
-        });
-        self.send_notification("edit", &edit_params);
-    }
-
-    fn file_open(&self, hwnd_owner: HWND) {
-        let filename = unsafe { get_open_file_dialog_path(hwnd_owner) };
-        if let Some(filename) = filename {
-            self.req_new_view(Some(&filename));
-            let mut state = self.state.borrow_mut();
-            state.edit_view.filename = Some(filename);
-            state.edit_view.clear_line_cache();
+impl AppState {
+    fn new() -> AppState {
+        AppState {
+            focused: Default::default(),
```

Comment:
This seems a little dicey. Maybe focused should be an `Option<ViewId>`? Of course, empty string can be interpreted to have the same meaning.

Reply:
Yes, I think you're right. Better to be explicit.